File: leetcode/dp/3466_Maximum_Coin_Collection/codes/dp.c
#include <stdlib.h>
#include <limits.h>

long long maxCoins(int* lane1, int lane1Size, int* lane2, int lane2Size) {
    int n = lane1Size;
    long long NEG = LLONG_MIN / 4;
    long long dpPrev[2][3];
    for (int l = 0; l < 2; l++) {
        for (int s = 0; s < 3; s++) {
            dpPrev[l][s] = NEG;
        }
    }
    long long ans = NEG;
    for (int i = 0; i < n; i++) {
        long long dpCurr[2][3];
        for (int l = 0; l < 2; l++) {
            for (int s = 0; s < 3; s++) {
                dpCurr[l][s] = NEG;
            }
        }
        for (int l = 0; l < 2; l++) {
            int gain = (l == 0? lane1[i] : lane2[i]);
            for (int s = 0; s < 3; s++) {
                long long best = NEG;
                int startS = (l == 0 ? 0 : 1);
                if (s == startS) {
                    if ((long long)gain > best) best = (long long)gain;
                }
                if (dpPrev[l][s] != NEG) {
                    long long cand = dpPrev[l][s] + (long long)gain;
                    if (cand > best) best = cand;
                }
                if (s >= 1 && dpPrev[1 - l][s - 1] != NEG) {
                    long long cand = dpPrev[1 - l][s - 1] + (long long)gain;
                    if (cand > best) best = cand;
                }
                dpCurr[l][s] = best;
                if (best != NEG && best > ans) ans = best;
            }
        }
        for (int l = 0; l < 2; l++) {
            for (int s = 0; s < 3; s++) {
                dpPrev[l][s] = dpCurr[l][s];
            }
        }
    }
    return ans;
}

File: leetcode/dp/3466_Maximum_Coin_Collection/codes/dp.py
from typing import List

class Solution:
    def maxCoins(self, lane1: List[int], lane2: List[int]) -> int:
        n = len(lane1)
        NEG =  - 10 ** 30
        dp_prev = [[NEG] * 3 for _ in range(2)]
        ans = NEG
        for i in range(n):
            dp_curr = [[NEG] * 3 for _ in range(2)]
            for l in range(2):
                arr = lane1 if l == 0 else lane2
                for s in range(3):
                    best = NEG
                    start_s = 0 if l == 0 else 1
                    if s == start_s:
                        best = max(best, arr[i])
                    if dp_prev[l][s] != NEG:
                        best = max(best, dp_prev[l][s] + arr[i])
                    if s >= 1 and dp_prev[1 - l][s - 1] != NEG:
                        best = max(best, dp_prev[1 - l][s - 1] + arr[i])
                    dp_curr[l][s] = best
                    if best != NEG and best > ans:
                        ans = best
            dp_prev = dp_curr
        return ans
